### PR TITLE
Add initial ADR records

### PR DESCRIPTION
**Pull request by neilbmclaughlin**

### CODE REVIEW

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

File: doc/architecture/decisions/0002-use-elastic-co-docker-image.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+Accepted
+
+## Context
+
+[ElasticSearch](https://www.elastic.co/products/elasticsearch) is a Java based search engine built on top of technologies like Lucene and Solr. It is open source software which can be freely [downloaded](https://github.com/elastic/elasticsearch) and installed on a number of operating systems. It is also available from elastic.co as both a Docker image and as a cloud service.
+
+## Decision
+
+For initial proof of concept and early beta the Docker image is the quickest way of getting up and running with ES and this outweighs the concerns about the inability to change the default user/password for the user which ES runs as.
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0002-use-elastic-co-docker-image.md]: The latter part of this sentence seems more like a consequence to me. Perhaps rewrite to `We will use the official Docker image provided by elastic.co`. I don't know if the bit about initial POCs and early Beta is required, especially as these things tend to no just be for that initial period.

File: doc/architecture/decisions/0002-use-elastic-co-docker-image.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+X-Pack: the included basic subscription for the included X-Pack security module does not allow the default username and passwords to be changed without paying a subscription. This will need addressing going forward and this could be by paying the X-Pack subscription, building our own ES image with a free security alternative (e.g. [SearchGuard](https://sematext.com/blog/2017/05/22/elasticsearch-kibana-security-search-guard/)) or by using the ES SaaS [offering](https://www.elastic.co/cloud/as-a-service).  
+
+The inability to change default user presents a small risk. The risk is small since 1) the instance of ES will not be public facing, 2) it can be secured at the Docker container and network level and 3) it contains only data which is currently in the public domain. It would in theory be possible, if a public facing container based app in the same virual network was compromised, for the data to be modified or deleted. The likelyhood of this happening is small.
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0002-use-elastic-co-docker-image.md]: I think the point above about having to use X-Pack to change the password is the thing that needs recording. I think this line i.e. line 23 can be removed.

File: doc/architecture/decisions/0002-use-elastic-co-docker-image.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+## Consequences
+
+Whilst the docker image is freely available there have been concerns raised in the ES community regarding it no longer being hosted on Docker Hub (it is now hosted on the elastic.co Docker [repo](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html) ). The Docker Hub [repo](https://hub.docker.com/_/elasticsearch/) has many comments and these can be distilled down to difficulty managing different versions of ES when not hosted on docker hub and the tie in with X-pack security features.
+
+X-Pack: the included basic subscription for the included X-Pack security module does not allow the default username and passwords to be changed without paying a subscription. This will need addressing going forward and this could be by paying the X-Pack subscription, building our own ES image with a free security alternative (e.g. [SearchGuard](https://sematext.com/blog/2017/05/22/elasticsearch-kibana-security-search-guard/)) or by using the ES SaaS [offering](https://www.elastic.co/cloud/as-a-service).
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0002-use-elastic-co-docker-image.md]: Suggest changing `going forward` to `before the service is presented to users by default` (or some other time we deem it needs to be addressed by).

File: doc/architecture/decisions/0002-use-elastic-co-docker-image.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+[ElasticSearch](https://www.elastic.co/products/elasticsearch) is a Java based search engine built on top of technologies like Lucene and Solr. It is open source software which can be freely [downloaded](https://github.com/elastic/elasticsearch) and installed on a number of operating systems. It is also available from elastic.co as both a Docker image and as a cloud service.
+
+## Decision
+
+For initial proof of concept and early beta the Docker image is the quickest way of getting up and running with ES and this outweighs the concerns about the inability to change the default user/password for the user which ES runs as.
+
+## Consequences
+
+Whilst the docker image is freely available there have been concerns raised in the ES community regarding it no longer being hosted on Docker Hub (it is now hosted on the elastic.co Docker [repo](https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html) ). The Docker Hub [repo](https://hub.docker.com/_/elasticsearch/) has many comments and these can be distilled down to difficulty managing different versions of ES when not hosted on docker hub and the tie in with X-pack security features.
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0002-use-elastic-co-docker-image.md]: Perhaps some rewording here? It sounds like the consequence of using the elastic.co image is that there are additional challenges in managing different versions of ES compared to using the Docker hub version.

Is that actually the case? Certainly the knock on effect of having images with X-Pack baked in is a consequence that needs recording (which is done below).

File: doc/architecture/decisions/0003-bulk-load-gp-data.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+Date: 2017-06-15
+
+## Status
+
+Accepted
+
+## Context
+
+In the early stages of ES adoption (which predates the development of ETL processes to automatically build the GP JSON data file) the GP data mas manually added to this repo. The data is in a format optimised for loading into MongoDB. Whilst Elastic Search (ES) can import documents singly in the format it takes some considerable time. By re-shaping the data to that required by the bulk import API of ES then the import can happen in a matter of seconds rather than 10's of minutes otherwise.
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0003-bulk-load-gp-data.md]: I think the context can be simplified quite a bit to something like `Other work within the same domain has created a JSON file containing all GP practices that need to be searched and exposed via ES. This datafile was originally created to be loaded into MongoDB.` (or something equally or more succinct.

File: doc/architecture/decisions/0003-bulk-load-gp-data.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+## Decision
+
+The use of an existing data file coupled with pre-procesing (using a `jq` script) to reshape it and subsequent bulk loading it into ES was a pragmatic move
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0003-bulk-load-gp-data.md]: Suggest rewording to `We will use the existing data file but reshape the data specifically to suit the needs of ES via jq (a lightweight JSON processing cli tool).` (or something similar).

File: doc/architecture/decisions/0003-bulk-load-gp-data.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ 
+## Consequences
+
+The data is loaded using bash scripts run within the container as part of `docker build` which means that the resulting Docker image contains ES preloaded with the GP data.
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0003-bulk-load-gp-data.md]: I don't think this needs to be here.

File: doc/architecture/decisions/0003-bulk-load-gp-data.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+The updating of the source GP data file in this repo is a manual process with no set schedule. This will likely be superceeded by the GP data ETL processors in the near future.
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0003-bulk-load-gp-data.md]: 👍

File: doc/architecture/decisions/0003-bulk-load-gp-data.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+As well as reshaping the GP JSON document for bulk insert some manipulation of the properties was necessary to optimise for particular search requirements. The reshaping was carried out using `jq` as a lightweight way of doing this but did mean that there are no tests around the reshaping.
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0003-bulk-load-gp-data.md]: 👍

**Review by st3v3nhunt — CHANGES_REQUESTED** [review]
A few comments

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by neilbmclaughlin** [comment]
@st3v3nhunt - I've reworked the ADR's in light of your comments and a re-reading of the Michael Nygard post. I've also split the decision record around usage of the ES Docker image into one for Docker usage in general and one for using the ES provided one specifically.

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

File: doc/architecture/decisions/0004-bulk-load-gp-data.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
+# 3. Build Image with GP Data
```
Comment (st3v3nhunt) [inline doc/architecture/decisions/0004-bulk-load-gp-data.md]: 3 should be a 4

**Review by st3v3nhunt — CHANGES_REQUESTED** [review]
Looking good. Just a small typo I noticed.

File: doc/architecture/decisions/0004-bulk-load-gp-data.md (hunk 10 above)
Comment (neilbmclaughlin) [reply doc/architecture/decisions/0004-bulk-load-gp-data.md]: @st3v3nhunt fixed and ready for merge.

**Comment by c2s-dev** [comment]
:rocket: deployment of profiles-db-elastic succeeded (http://profiles-db-elastic-pr-32.dev.beta.nhschoices.net)

**Review by st3v3nhunt — APPROVED** [review]
